### PR TITLE
common: fs: Rework the Common Filesystem interface to make use of std::filesystem

### LINKED ISSUE #4876: SMO: Uses the base profile for saves regardless of selected profile
SMO ignores the selected profile in yuzu and uses the base profile for loading a save.

[yuzu_log.txt](https://github.com/yuzu-emu/yuzu/files/5468415/yuzu_log.txt)

### PR DESCRIPTION
**Pull request by Morph1984**
This PR reimplements and reworks the entire Common Filesystem interface to make use of C++17/20's new std::filesystem library, removing many unused functions and leftovers from Citra, with a few remaining functions kept for compatibility with the current VFS implementation.

This continues from @lioncash 's previous work at #5173, and fixes all the Windows-specific issues this had,

I do not expect a change in filesystem performance, but this brings us closer to how HOS' filesystem API works.

List of changes:
- Place all filesystem code in common/fs
- Split filesystem code into 3 distinct files, `file.cpp/h`, `fs.cpp/h`, `path_util.cpp/h`, with several supporting files `fs_util.cpp/h`, `fs_types.h`, `fs_paths.h`
- Documented all Common::FS interface functions
- Make use of `std::span` where applicable in `Common::FS::IOFile`
- Cleaned up `Common::FS::IOFile` implementation and added extra utility functions
- Deprecated old path utility functions used in the current VFS implementation.
- A LOT of refactoring anywhere the previous Common::FS implementation is used, fixing issues along the way.

Special Thanks to:
- @Its-Rei , @BSoDGamingYT for general regression testing
- @lat9nq for all the linux help and discovering this fixes #4876
- @gidoly for helping me test and discovering several issues on non-English system locales

Fixes #4876

### CODE REVIEW

File: src/common/fs/fs.h [hunk 1]
```diff
@@ -0,0 +1,558 @@
...
+ * Returns whether a filesystem object at path is a directory.
+ *
+ * @param path Filesystem path
+ *
+ * @returns True if a filesystem object at path is a directory, false otherwise.
+ */
+[[nodiscard]] bool IsDir(const std::filesystem::path& path);
+
+template <typename Path>
+[[nodiscard]] bool IsDir(Path path) {
```
Comment (lioncash) [inline src/common/fs/fs.h:449]:
```suggestion
[[nodiscard]] bool IsDir(const Path& path) {
```

File: src/common/fs/fs.h [hunk 2]
```diff
@@ -0,0 +1,558 @@
...
+ * Returns whether a filesystem object at path is a file.
+ *
+ * @param path Filesystem path
+ *
+ * @returns True if a filesystem object at path is a file, false otherwise.
+ */
+[[nodiscard]] bool IsFile(const std::filesystem::path& path);
+
+template <typename Path>
+[[nodiscard]] bool IsFile(Path path) {
```
Comment (lioncash) [inline src/common/fs/fs.h:430]:
```suggestion
[[nodiscard]] bool IsFile(const Path& path) {
```

File: src/common/fs/fs.h [hunk 3]
```diff
@@ -0,0 +1,558 @@
...
+ * Returns whether a filesystem object at path exists.
+ *
+ * @param path Filesystem path
+ *
+ * @returns True if a filesystem object at path exists, false otherwise.
+ */
+[[nodiscard]] bool Exists(const std::filesystem::path& path);
+
+template <typename Path>
+[[nodiscard]] bool Exists(Path path) {
```
Comment (lioncash) [inline src/common/fs/fs.h:411]:
```suggestion
[[nodiscard]] bool Exists(const Path& path) {
```

Ditto for other relevant occurrences above and below

File: src/common/fs/path_util.h [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+ * @param first First filesystem path
+ * @param second Second filesystem path
+ *
+ * @returns A concatenated filesystem path.
+ */
+[[nodiscard]] std::filesystem::path ConcatPath(const std::filesystem::path& first,
+                                               const std::filesystem::path& second);
+
+template <typename Path1, typename Path2>
+[[nodiscard]] std::filesystem::path ConcatPath(Path1 first, Path2 second) {
```
Comment (lioncash) [inline src/common/fs/path_util.h:65]:
```suggestion
[[nodiscard]] std::filesystem::path ConcatPath(const Path1& first, const Path2& second) {
```

Ditto elswhere this is applicable

File: src/common/fs/file.h [hunk 5]
```diff
@@ -0,0 +1,448 @@
...
+ * @param open_mode File stream open mode
+ */
+template <typename FileStream>
+void OpenFileStream(FileStream& file_stream, const std::filesystem::path& path,
+                    std::ios_base::openmode open_mode) {
+    file_stream.open(path, open_mode);
+}
+
+template <typename FileStream, typename Path>
+void OpenFileStream(FileStream& file_stream, Path path, std::ios_base::openmode open_mode) {
```
Comment (lioncash) [inline src/common/fs/file.h:43]:
```suggestion
void OpenFileStream(FileStream& file_stream, const Path& path, std::ios_base::openmode open_mode) {
```

Ditto where applicable in the file

**Review by lioncash — COMMENTED** [review]
Very well done. Review comments are just minor nitpicks

A+ documentation work on the added interfaces. Wish more of the codebase was this well documented.

**Comment by Kelebek1** [comment]
Mod files remain in-use by yuzu after closing the game they're a part of, when they should be cleaned up. I.e, replace any game file via mod, load that game, stop emulation, try to delete/rename the file and you'll find you can't as it's still in-use.

Could this be fixed as part of this PR, or is it outside the scope of this change?

**Comment by Morph1984** [comment]
> Mod files remain in-use by yuzu after closing the game they're a part of, when they should be cleaned up. I.e, replace any game file via mod, load that game, stop emulation, try to delete/rename the file and you'll find you can't as it's still in-use. Yuzu will crash on open (not even loading a game, just trying to open it) if you try to open a second instance after stopping emulation, giving OpenFile permission denied errors.
> 
> Ideally the file would only be read and then closed when needed by the game, and not kept open for yuzu's entire lifespan, but just closing all files on stopping emulation would be beneficial.
> 
> Could this be fixed as part of this PR, or is it outside the scope of this change? I could make a general issue for it if so.

This is unfortunately not within the scope of the PR as the main goal is simply to port over the current interface to the new interface. I can look at the issue more closely after this is merged.

File: src/video_core/vulkan_common/nsight_aftermath_tracker.cpp [hunk 6]
```diff
@@ -161,15 +169,16 @@ void NsightAftermathTracker::OnShaderDebugInfoCallback(const void* shader_debug_
...
 
-    const std::string path =
-        fmt::format("{}/shader_{:016x}{:016x}.nvdbg", dump_dir, identifier.id[0], identifier.id[1]);
-    Common::FS::IOFile file(path, "wb");
+    const auto path =
+        dump_dir / fmt::format("shader_{:016x}{:016x}.nvdbg", identifier.id[0], identifier.id[1]);
+    Common::FS::IOFile file{path, Common::FS::FileAccessMode::Write,
+                            Common::FS::FileType::BinaryFile};
     if (!file.IsOpen()) {
         LOG_ERROR(Render_Vulkan, "Failed to create file {}", path);
```
Comment (ameerj) [inline src/video_core/vulkan_common/nsight_aftermath_tracker.cpp:177]: fmt can't format `path` when compiling on windows.

File: src/video_core/vulkan_common/nsight_aftermath_tracker.cpp [hunk 7]
```diff
@@ -161,15 +169,16 @@ void NsightAftermathTracker::OnShaderDebugInfoCallback(const void* shader_debug_
...
         return;
     }
-    if (file.WriteBytes(static_cast<const u8*>(shader_debug_info), shader_debug_info_size) !=
-        shader_debug_info_size) {
+    if (file.WriteSpan(std::span(static_cast<const u8*>(shader_debug_info),
+                                 shader_debug_info_size)) != shader_debug_info_size) {
         LOG_ERROR(Render_Vulkan, "Failed to write file {}", path);
```
Comment (ameerj) [inline src/video_core/vulkan_common/nsight_aftermath_tracker.cpp:182]: fmt can't format `path` when compiling on windows.

File: src/core/file_sys/vfs_real.cpp [hunk 8]
```diff
@@ -462,14 +459,18 @@ std::map<std::string, VfsEntryType, std::less<>> RealVfsDirectory::GetEntries()
...
+                              out.insert_or_assign(filename, FS::IsDir(full_path)
+                                                                 ? VfsEntryType::Directory
+                                                                 : VfsEntryType::File);
+
+                              return true;
+                          }
+
+    );
 
     return out;
```
Comment (ameerj) [inline src/core/file_sys/vfs_real.cpp:475]:
```suggestion
    const FS::DirEntryCallable callable = [&out](const std::filesystem::path& full_path) {
        const auto filename = FS::PathToUTF8String(full_path.filename());
        const auto type = FS::IsDir(full_path) ? VfsEntryType::Directory : VfsEntryType::File;
        out.insert_or_assign(filename, type);
        return true;
    };

    FS::IterateDirEntries(path, callable);
    return out;
```

File: src/core/perf_stats.cpp [hunk 9]
```diff
@@ -38,12 +40,17 @@ PerfStats::~PerfStats() {
...
-    const std::string& path = Common::FS::GetUserPath(Common::FS::UserPath::LogDir);
+
+    const auto path = Common::FS::GetYuzuPath(Common::FS::YuzuPath::LogDir);
     // %F Date format expanded is "%Y-%m-%d"
-    const std::string filename =
-        fmt::format("{}/{:%F-%H-%M}_{:016X}.csv", path, *std::localtime(&t), title_id);
-    Common::FS::IOFile file(filename, "w");
-    file.WriteString(stream.str());
+    const auto filename =
+        path / fmt::format("{:%F-%H-%M}_{:016X}.csv", *std::localtime(&t), title_id);
```
Comment (ameerj) [inline src/core/perf_stats.cpp:47]: makes it look more tidy, and more accurate variable names.
```suggestion
    const auto filename = fmt::format("{:%F-%H-%M}_{:016X}.csv", *std::localtime(&t), title_id);
    const auto filepath = path / filename;
```

File: src/video_core/renderer_opengl/gl_shader_disk_cache.cpp [hunk 10]
```diff
@@ -26,11 +27,7 @@ using Tegra::Engines::ShaderType;
 using VideoCommon::Shader::BindlessSamplerMap;
 using VideoCommon::Shader::BoundSamplerMap;
 using VideoCommon::Shader::KeyMap;
-
-namespace {
```
Comment (ameerj) [inline src/video_core/renderer_opengl/gl_shader_disk_cache.cpp:30]: Why did you move this?

File: src/video_core/renderer_opengl/gl_shader_disk_cache.cpp [hunk 11]
```diff
@@ -116,27 +119,29 @@ bool ShaderDiskCacheEntry::Load(Common::FS::IOFile& file) {
...
-        file.ReadArray(flat_bindless_samplers.data(), flat_bindless_samplers.size()) !=
-            flat_bindless_samplers.size()) {
+
+    if (file.Read(flat_keys) != flat_keys.size() ||
+        file.Read(flat_bound_samplers) != flat_bound_samplers.size() ||
+        file.Read(flat_separate_samplers) != flat_separate_samplers.size() ||
+        file.Read(flat_bindless_samplers) != flat_bindless_samplers.size()) {
         return false;
     }
+
```
Comment (ameerj) [inline src/video_core/renderer_opengl/gl_shader_disk_cache.cpp:129]: Unnecessary newlines

File: src/video_core/renderer_opengl/gl_shader_disk_cache.cpp [hunk 12]
```diff
@@ -116,27 +119,29 @@ bool ShaderDiskCacheEntry::Load(Common::FS::IOFile& file) {
...
     for (const auto& entry : flat_keys) {
         keys.insert({{entry.cbuf, entry.offset}, entry.value});
     }
+
     for (const auto& entry : flat_bound_samplers) {
         bound_samplers.emplace(entry.offset, entry.sampler);
     }
+
```
Comment (ameerj) [inline src/video_core/renderer_opengl/gl_shader_disk_cache.cpp:137]: Unnecessary newlines

File: src/video_core/renderer_opengl/gl_shader_disk_cache.cpp [hunk 13]
```diff
@@ -349,9 +363,11 @@ void ShaderDiskCacheOpenGL::SaveEntry(const ShaderDiskCacheEntry& entry) {
     }
 
     Common::FS::IOFile file = AppendTransferableFile();
+
```
Comment (ameerj) [inline src/video_core/renderer_opengl/gl_shader_disk_cache.cpp:366]: Unnecessary newlines

File: src/video_core/renderer_opengl/gl_shader_disk_cache.cpp [hunk 14]
```diff
@@ -349,9 +363,11 @@ void ShaderDiskCacheOpenGL::SaveEntry(const ShaderDiskCacheEntry& entry) {
...
     if (!file.IsOpen()) {
         return;
     }
+
```
Comment (ameerj) [inline src/video_core/renderer_opengl/gl_shader_disk_cache.cpp:370]: Unnecessary newlines

File: src/video_core/renderer_opengl/gl_shader_disk_cache.cpp [hunk 15]
```diff
@@ -395,22 +411,27 @@ Common::FS::IOFile ShaderDiskCacheOpenGL::AppendTransferableFile() const {
         return {};
     }
 
-    const auto transferable_path{GetTransferablePath()};
+    const auto transferable_path = GetTransferablePath();
```
Comment (ameerj) [inline src/video_core/renderer_opengl/gl_shader_disk_cache.cpp:414]: Don't change the style of code that you are not rewriting

File: src/yuzu/game_list_worker.cpp [hunk 16]
```diff
@@ -70,12 +71,14 @@ std::pair<std::vector<u8>, std::string> GetGameListCachedObject(
...
-    const auto path1 = Common::FS::GetUserPath(Common::FS::UserPath::CacheDir) + DIR_SEP +
-                       "game_list" + DIR_SEP + filename + ".jpeg";
-    const auto path2 = Common::FS::GetUserPath(Common::FS::UserPath::CacheDir) + DIR_SEP +
-                       "game_list" + DIR_SEP + filename + ".appname.txt";
+    const auto path1 =
+        Common::FS::PathToUTF8String(Common::FS::GetYuzuPath(Common::FS::YuzuPath::CacheDir) /
+                                     "game_list" / fmt::format("{}.jpeg", filename));
+    const auto path2 =
+        Common::FS::PathToUTF8String(Common::FS::GetYuzuPath(Common::FS::YuzuPath::CacheDir) /
+                                     "game_list" / fmt::format("{}.appname.txt", filename));
```
Comment (ameerj) [inline src/yuzu/game_list_worker.cpp:79]:
```suggestion
    const auto cache_dir = Common::FS::GetYuzuPath(Common::FS::YuzuPath::CacheDir) / "game_list";
    const auto jpeg_name = fmt::format("{}.jpeg", filename);
    const auto txt_name = fmt::format("{}.appname.txt", filename);
    const auto path1 = Common::FS::PathToUTF8String(cache_dir / jpeg_name);
    const auto path2 = Common::FS::PathToUTF8String(cache_dir / txt_name);
```

File: src/yuzu/main.cpp [hunk 17]
```diff
@@ -178,36 +179,27 @@ static void InitializeLogging() {
...
-    Common::FS::DeleteDirRecursively(offline_legal_information);
-    Common::FS::DeleteDirRecursively(offline_system_data);
+    const auto offline_fonts = Common::FS::GetYuzuPath(Common::FS::YuzuPath::CacheDir) / "fonts";
+    const auto offline_manual =
+        Common::FS::GetYuzuPath(Common::FS::YuzuPath::CacheDir) / "offline_web_applet_manual";
+    const auto offline_legal_information = Common::FS::GetYuzuPath(Common::FS::YuzuPath::CacheDir) /
+                                           "offline_web_applet_legal_information";
+    const auto offline_system_data =
+        Common::FS::GetYuzuPath(Common::FS::YuzuPath::CacheDir) / "offline_web_applet_system_data";
+
```
Comment (ameerj) [inline src/yuzu/main.cpp:198]:
```suggestion
    const auto cache_dir = Common::FS::GetYuzuPath(Common::FS::YuzuPath::CacheDir);
    const auto offline_fonts = cache_dir / "fonts";
    const auto offline_manual = cache_dir / "offline_web_applet_manual";
    const auto offline_legal_information = cache_dir / "offline_web_applet_legal_information";
    const auto offline_system_data = cache_dir / "offline_web_applet_system_data";
```

File: src/yuzu/main.cpp [hunk 18]
```diff
@@ -1579,34 +1572,35 @@ void GMainWindow::OnGameListOpenFolder(u64 program_id, GameListOpenTarget target
             Service::Account::ProfileManager manager;
             const auto user_id = manager.GetUser(static_cast<std::size_t>(index));
             ASSERT(user_id);
-            path = nand_dir + FileSys::SaveDataFactory::GetFullPath(
-                                  system, FileSys::SaveDataSpaceId::NandUser,
-                                  FileSys::SaveDataType::SaveData, program_id, user_id->uuid, 0);
+            path = Common::FS::ConcatPathSafe(
```
Comment (ameerj) [inline src/yuzu/main.cpp:1575]: Follow similar formatting cleanup as before

File: src/yuzu/main.cpp [hunk 19]
```diff
@@ -1886,8 +1875,9 @@ void GMainWindow::OnGameListDumpRomFS(u64 program_id, const std::string& game_pa
         return;
     }
 
-    const auto path = fmt::format(
-        "{}{:016X}/romfs", Common::FS::GetUserPath(Common::FS::UserPath::DumpDir), *romfs_title_id);
+    const auto path =
```
Comment (ameerj) [inline src/yuzu/main.cpp:1878]: ditto

File: src/yuzu/main.cpp [hunk 20]
```diff
@@ -1813,21 +1809,17 @@ void GMainWindow::OnGameListRemoveFile(u64 program_id, GameListRemoveTarget targ
...
-    const QString shader_dir =
-        QString::fromStdString(Common::FS::GetUserPath(Common::FS::UserPath::ShaderDir));
-    const QString transferable_shader_cache_folder_path =
-        shader_dir + QStringLiteral("opengl") + QDir::separator() + QStringLiteral("transferable");
-    const QString transferable_shader_cache_file_path =
-        transferable_shader_cache_folder_path + QDir::separator() +
-        QString::fromStdString(fmt::format("{:016X}.bin", program_id));
-
-    if (!QFile::exists(transferable_shader_cache_file_path)) {
+    const auto transferable_shader_cache_file_path =
```
Comment (ameerj) [inline src/yuzu/main.cpp:1812]: ditto

File: src/yuzu/main.cpp [hunk 21]
```diff
@@ -3265,12 +3262,18 @@ int main(int argc, char* argv[]) {
     QCoreApplication::setOrganizationName(QStringLiteral("yuzu team"));
     QCoreApplication::setApplicationName(QStringLiteral("yuzu"));
 
+#ifdef _WIN32
```
Comment (ameerj) [inline src/yuzu/main.cpp:3265]: unnecessary newlines ditto

File: src/yuzu_cmd/config.cpp [hunk 22]
```diff
@@ -30,29 +32,32 @@ namespace FS = Common::FS;
...
-            FS::WriteStringToFile(true, location, default_contents);
-            sdl2_config = std::make_unique<INIReader>(location); // Reopen file
+            LOG_WARNING(Config, "Failed to load {}. Creating file from defaults...",
+                        FS::PathToUTF8String(sdl2_config_loc));
+
+            void(FS::CreateParentDir(sdl2_config_loc));
+            void(FS::WriteStringToFile(sdl2_config_loc, FS::FileType::TextFile, default_contents));
+
+            sdl2_config =
+                std::make_unique<INIReader>(FS::PathToUTF8String(sdl2_config_loc)); // Reopen file
```
Comment (ameerj) [inline src/yuzu_cmd/config.cpp:53]: Code comments go above the line. although this comment doesn't seem too accurate/helpful anymore.
```suggestion
            // Reopen file
            sdl2_config = std::make_unique<INIReader>(FS::PathToUTF8String(sdl2_config_loc));
```

File: src/core/file_sys/vfs_real.cpp [hunk 23]
```diff
@@ -94,25 +94,29 @@ VirtualFile RealVfsFilesystem::OpenFile(std::string_view path_, Mode perms) {
...
 VirtualFile RealVfsFilesystem::CopyFile(std::string_view old_path_, std::string_view new_path_) {
-    const auto old_path = FS::SanitizePath(old_path_, FS::DirectorySeparator::PlatformDefault);
-    const auto new_path = FS::SanitizePath(new_path_, FS::DirectorySeparator::PlatformDefault);
-
-    if (!FS::Exists(old_path) || FS::Exists(new_path) || FS::IsDirectory(old_path) ||
-        !FS::Copy(old_path, new_path)) {
-        return nullptr;
-    }
-    return OpenFile(new_path, Mode::ReadWrite);
+    // Unused
```
[the remark this answers is not in the document]
Comment (Morph1984) [reply src/core/file_sys/vfs_real.cpp:118]: Since this inherits from a virtual base class, this has to be stubbed

File: src/core/file_sys/vfs_real.cpp [hunk 24]
```diff
@@ -168,35 +172,25 @@ VirtualDir RealVfsFilesystem::OpenDirectory(std::string_view path_, Mode perms)
...
 VirtualDir RealVfsFilesystem::CopyDirectory(std::string_view old_path_,
                                             std::string_view new_path_) {
-    const auto old_path = FS::SanitizePath(old_path_, FS::DirectorySeparator::PlatformDefault);
-    const auto new_path = FS::SanitizePath(new_path_, FS::DirectorySeparator::PlatformDefault);
-    if (!FS::Exists(old_path) || FS::Exists(new_path) || !FS::IsDirectory(old_path)) {
-        return nullptr;
-    }
-    FS::CopyDir(old_path, new_path);
-    return OpenDirectory(new_path, Mode::ReadWrite);
+    // Unused
```
[the remark this answers is not in the document]
Comment (Morph1984) [reply src/core/file_sys/vfs_real.cpp:184]: Since this inherits from a virtual base class, this has to be stubbed

File: src/core/file_sys/vfs_real.cpp [hunk 25]
```diff
@@ -266,41 +259,41 @@ std::size_t RealVfsFile::GetSize() const {
...
 
 bool RealVfsFile::Rename(std::string_view name) {
-    return base.MoveFile(path, parent_path + DIR_SEP + std::string(name)) != nullptr;
+    return base.MoveFile(path, parent_path + '/' + std::string(name)) != nullptr;
 }
 
-bool RealVfsFile::Close() {
-    return backing->Close();
+void RealVfsFile::Close() {
+    backing->Close();
```
[the remark this answers is not in the document]
Comment (Morph1984) [reply src/core/file_sys/vfs_real.cpp:296]: File closure is not checked anywhere

File: src/video_core/renderer_opengl/gl_shader_disk_cache.cpp (hunk 10 above)
Comment (Morph1984) [reply src/video_core/renderer_opengl/gl_shader_disk_cache.cpp:30]: To ensure that the underlying type can be deduced by Read

File: src/core/file_sys/vfs_real.cpp [hunk 26]
```diff
@@ -81,11 +80,12 @@ VirtualFile RealVfsFilesystem::OpenFile(std::string_view path_, Mode perms) {
         }
     }
 
-    if (!FS::Exists(path) && True(perms & Mode::WriteAppend)) {
-        FS::CreateEmptyFile(path);
+    auto backing = FS::FileOpen(path, ModeFlagsToFileAccessMode(perms), FS::FileType::BinaryFile);
```
[the remark this answers is not in the document]
Comment (Morph1984) [reply src/core/file_sys/vfs_real.cpp:83]: backing is `std::move`d into the map, this will result into a copy

File: src/common/fs/file.cpp [hunk 27]
```diff
@@ -0,0 +1,409 @@
...
+            return "rb";
+        case FileAccessMode::Write:
+            return "wb";
+        case FileAccessMode::Append:
+            return "ab";
+        case FileAccessMode::ReadWrite:
+            return "r+b";
+        case FileAccessMode::ReadAppend:
+            return "a+b";
+        default:
```
[the remark this answers is not in the document]
Comment (Morph1984) [reply src/common/fs/file.cpp]: Not handling the default case will result into a compilation error

File: src/core/file_sys/vfs_real.cpp [hunk 28]
```diff
@@ -7,42 +7,41 @@
...
+        return FS::FileAccessMode::ReadWrite;
+    case Mode::Append:
+        return FS::FileAccessMode::Append;
+    case Mode::ReadAppend:
+        return FS::FileAccessMode::ReadAppend;
+    case Mode::WriteAppend:
+        return FS::FileAccessMode::Append;
+    case Mode::All:
+        return FS::FileAccessMode::ReadAppend;
+    default:
```
[the remark this answers is not in the document]
Comment (Morph1984) [reply src/core/file_sys/vfs_real.cpp:38]: Not handling the default case will result into a compilation error

File: src/core/file_sys/vfs_real.cpp [hunk 29]
```diff
@@ -208,7 +202,7 @@ VirtualDir RealVfsFilesystem::MoveDirectory(std::string_view old_path_,
 
         const auto file_old_path =
             FS::SanitizePath(kv.first, FS::DirectorySeparator::PlatformDefault);
-        auto file_new_path = FS::SanitizePath(new_path + DIR_SEP + kv.first.substr(old_path.size()),
+        auto file_new_path = FS::SanitizePath(new_path + '/' + kv.first.substr(old_path.size()),
```
[the remark this answers is not in the document]
Comment (Morph1984) [reply src/core/file_sys/vfs_real.cpp:205]: file_new_path is `std::move`d into the map a few lines down

File: src/core/file_sys/vfs_real.cpp (hunk 28 above)
[the remark this answers is not in the document]
Comment (ameerj) [reply src/core/file_sys/vfs_real.cpp:38]: That's the intention. I would rather each mode have an explicit case accounted for. A precautionary measure against modifications to the enum without specifying the way it will be used.

File: src/yuzu_cmd/config.cpp [hunk 30]
```diff
@@ -30,29 +32,31 @@ namespace FS = Common::FS;
...
-            FS::CreateFullPath(location);
-            FS::WriteStringToFile(true, location, default_contents);
-            sdl2_config = std::make_unique<INIReader>(location); // Reopen file
+            LOG_WARNING(Config, "Failed to load {}. Creating file from defaults...",
+                        FS::PathToUTF8String(sdl2_config_loc));
+
+            void(FS::CreateParentDir(sdl2_config_loc));
+            void(FS::WriteStringToFile(sdl2_config_loc, FS::FileType::TextFile, default_contents));
+
+            sdl2_config = std::make_unique<INIReader>(FS::PathToUTF8String(sdl2_config_loc));
```
Comment (ameerj) [inline src/yuzu_cmd/config.cpp:52]:
```suggestion
        const auto config_loc_str = FS::PathToUTF8String(sdl2_config_loc);
        LOG_WARNING(Config, "Failed to load {}. Creating file from defaults...", config_loc_str);
        void(FS::CreateParentDir(sdl2_config_loc));
        void(FS::WriteStringToFile(sdl2_config_loc, FS::FileType::TextFile, default_contents));
        sdl2_config = std::make_unique<INIReader>(config_loc_str);
```

File: src/yuzu/main.cpp [hunk 31]
```diff
@@ -2936,18 +2936,17 @@ void GMainWindow::OnReinitializeKeys(ReinitializeKeyBehavior behavior) {
...
-        Common::FS::Delete(Common::FS::GetUserPath(Common::FS::UserPath::KeysDir) +
-                           "console.keys_autogenerated");
-        Common::FS::Delete(Common::FS::GetUserPath(Common::FS::UserPath::KeysDir) +
-                           "title.keys_autogenerated");
+        void(Common::FS::RemoveFile(Common::FS::GetYuzuPath(Common::FS::YuzuPath::KeysDir) /
+                                    "prod.keys_autogenerated"));
+        void(Common::FS::RemoveFile(Common::FS::GetYuzuPath(Common::FS::YuzuPath::KeysDir) /
+                                    "console.keys_autogenerated"));
+        void(Common::FS::RemoveFile(Common::FS::GetYuzuPath(Common::FS::YuzuPath::KeysDir) /
+                                    "title.keys_autogenerated"));
```
Comment (ameerj) [inline src/yuzu/main.cpp:2944]:
```suggestion
        const auto keys_dir = Common::FS::GetYuzuPath(Common::FS::YuzuPath::KeysDir);
        void(Common::FS::RemoveFile(keys_dir / "prod.keys_autogenerated"));
        void(Common::FS::RemoveFile(keys_dir / "console.keys_autogenerated"));
        void(Common::FS::RemoveFile(keys_dir / "title.keys_autogenerated"));
```

File: src/yuzu/main.cpp [hunk 32]
```diff
@@ -1965,24 +1957,29 @@ void GMainWindow::OnGameListNavigateToGamedbEntry(u64 program_id,
...
-        path = directory;
+        fs_path = directory.toStdString();
     }
-    if (!QFileInfo::exists(path)) {
-        QMessageBox::critical(this, tr("Error Opening %1").arg(path), tr("Folder does not exist!"));
+
+    const auto qt_path = QString::fromStdString(Common::FS::PathToUTF8String(fs_path));
+
+    if (!Common::FS::IsDir(fs_path)) {
+        QMessageBox::critical(this, tr("Error Opening %1").arg(qt_path),
```
Comment (ameerj) [inline src/yuzu/main.cpp:1977]: this is checking `fs_path` but the error prints `qt_path`

File: src/yuzu/main.cpp [hunk 33]
```diff
@@ -1965,24 +1957,29 @@ void GMainWindow::OnGameListNavigateToGamedbEntry(u64 program_id,
...
+        fs_path =
+            Common::FS::GetYuzuPath(Common::FS::YuzuPath::NANDDir) / "system/Contents/registered";
     } else {
-        path = directory;
+        fs_path = directory.toStdString();
     }
-    if (!QFileInfo::exists(path)) {
-        QMessageBox::critical(this, tr("Error Opening %1").arg(path), tr("Folder does not exist!"));
+
+    const auto qt_path = QString::fromStdString(Common::FS::PathToUTF8String(fs_path));
```
Comment (ameerj) [inline src/yuzu/main.cpp:1974]: `qt_path` is instantiated before checking if `fs_path` is valid.

File: src/yuzu/game_list_worker.cpp [hunk 34]
```diff
@@ -281,23 +285,28 @@ void GameListWorker::AddTitlesToGameList(GameListDir* parent_dir) {
...
             // Breaks the callback loop.
             return false;
         }
 
-        const std::string physical_name = directory + DIR_SEP + virtual_name;
-        const bool is_dir = Common::FS::IsDirectory(physical_name);
+        const auto physical_name = Common::FS::PathToUTF8String(path);
+
+        const auto is_dir = Common::FS::IsDir(path);
+
```
Comment (ameerj) [inline src/yuzu/game_list_worker.cpp:302]: awkward newlines here.

File: src/yuzu/configuration/config.cpp [hunk 35]
```diff
@@ -243,27 +243,28 @@ const std::array<UISettings::Shortcut, 17> Config::default_hotkeys{{
 // clang-format on
 
 void Config::Initialize(const std::string& config_name) {
+    const auto fs_config_loc = FS::GetYuzuPath(FS::YuzuPath::ConfigDir);
```
Comment (ameerj) [inline src/yuzu/configuration/config.cpp:246]: add `auto config_file = fmt::format("{}.ini", config_name);` and reuse `config_file` where applicable here

File: src/core/hle/service/filesystem/filesystem.cpp [hunk 36]
```diff
@@ -728,14 +728,14 @@ void FileSystemController::CreateFactories(FileSys::VfsFilesystem& vfs, bool ove
...
-    auto dump_directory = vfs.OpenDirectory(Common::FS::GetUserPath(Common::FS::UserPath::DumpDir),
-                                            FileSys::Mode::ReadWrite);
+    auto nand_directory = vfs.OpenDirectory(
+        Common::FS::GetYuzuPathString(Common::FS::YuzuPath::NANDDir), FileSys::Mode::ReadWrite);
+    auto sd_directory = vfs.OpenDirectory(
+        Common::FS::GetYuzuPathString(Common::FS::YuzuPath::SDMCDir), FileSys::Mode::ReadWrite);
+    auto load_directory = vfs.OpenDirectory(
+        Common::FS::GetYuzuPathString(Common::FS::YuzuPath::LoadDir), FileSys::Mode::ReadWrite);
+    auto dump_directory = vfs.OpenDirectory(
+        Common::FS::GetYuzuPathString(Common::FS::YuzuPath::DumpDir), FileSys::Mode::ReadWrite);
```
Comment (ameerj) [inline src/core/hle/service/filesystem/filesystem.cpp:738]:
```suggestion
        using YuzuPath = Common::FS::YuzuPath;
        const auto rw_mode = FileSys::Mode::ReadWrite;
        auto nand_directory = vfs.OpenDirectory(Common::FS::GetYuzuPathString(YuzuPath::NANDDir), rw_mode);
        auto sd_directory = vfs.OpenDirectory(Common::FS::GetYuzuPathString(YuzuPath::SDMCDir), rw_mode);
        auto load_directory = vfs.OpenDirectory(Common::FS::GetYuzuPathString(YuzuPath::LoadDir), rw_mode);
        auto dump_directory = vfs.OpenDirectory(Common::FS::GetYuzuPathString(YuzuPath::DumpDir), rw_mode);
```
Or something to this effect, just to tidy this up a bit

File: src/common/fs/fs_types.h [hunk 37]
```diff
@@ -0,0 +1,77 @@
...
+
+#pragma once
+
+#include <functional>
+
+#include "common/common_funcs.h"
+#include "common/common_types.h"
+
+namespace Common::FS {
+
```
[the remark this answers is not in the document]
Comment (lioncash) [reply src/common/fs/fs_types.h:13]: This one's quite normal throughout the codebase. Some files have the extra newline here and some don't, so I think this spacing is just a personal preference thing.

Agree with the ones about spaces in preprocessor conditionals inside functions though. Though, when used outside of functions, I think the newline (on the outside of the preprocessor keywords) is nice as a way to make it stand out and call attention to itself a little more.

[the remark this answers is not in the document]
Comment (ameerj) [reply src/common/fs/fs_types.h:13]: I agree Lion, outside of a function it's preferred to follow the same newline padding around every function and preprocessor blocks. 
I was mainly pointing out that within the preprocessor conditional blocks there are unnecessary newlines (after the `#ifdef` and before the `#endif`)

File: src/common/fs/file.h [hunk 38]
```diff
@@ -0,0 +1,460 @@
...
+ * @param path Filesystem path
+ * @param open_mode File stream open mode
+ */
+template <typename FileStream>
+void OpenFileStream(FileStream& file_stream, const std::filesystem::path& path,
+                    std::ios_base::openmode open_mode) {
+    file_stream.open(path, open_mode);
+}
+
+#ifdef _WIN32
```
Comment (ameerj) [inline src/common/fs/file.h:39]: Seems that most of these functions need only be defined once for windows. Please review these platform dependent functions and follow the 
```
#ifdef _WIN32
...
#else
...
#endif
```
pattern.

Comment (ameerj) [reply src/common/fs/file.h:39]: In fact it's probably preferred to define _all_ windows specific functions into one large preprocessor condition, with all non-windows variants defined in the else block. similar to https://github.com/yuzu-emu/yuzu/blob/master/src/core/network/network.cpp

File: src/yuzu/main.cpp (hunk 32 above)
Comment (Morph1984) [reply src/yuzu/main.cpp:1978]: qt_path is simply the Qt string version of fs_path. this is intentional

File: src/yuzu/main.cpp (hunk 33 above)
Comment (Morph1984) [reply src/yuzu/main.cpp:1975]: We use qt_path in logging the error and below it

**Comment by lioncash** [comment]
Alright, where are we at with this? Is it good to go?

**Comment by ameerj** [comment]
> 
> 
> Alright, where are we at with this? Is it good to go?

There's the nit-pick of redundant newlines on the _inside_ of preprocessor conditionals still outstanding. Should be good otherwise 👍

**Comment by Morph1984** [comment]
@Kelebek1 The latest commit may fix the issue(s) you have encountered

**Comment by lioncash** [comment]
Alright what's the status now? Can I hit the green button?

**Review by ameerj — APPROVED** [review]
